<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="bioloid_without_legs">

    <!-- Pi is now predefined -->
    <!-- <xacro:property name="pi" value="3.14159"/> -->
    <xacro:property name="cad_dir" value="file://$(find bioloid_master)/../../assets/cad"/>
    <!-- Typically, STL/DAE model units are in mm, but RViz units are in m, hence models must be scaled down by 1000. -->
    <xacro:macro name="default_geometry" params="cad_file suffix:=_FCD-Conv_BDR-Col extension:=dae">
        <geometry>
            <mesh filename="${cad_dir}/${cad_file}${suffix}.${extension}" scale="0.001 0.001 0.001"/>
        </geometry>
    </xacro:macro>
    <xacro:macro name="z_rot">
        <axis xyz="0 0 1"/>
    </xacro:macro>

    <!-- Select one and only one of the following include files -->
    <!-- 1) Standard parts of Bioloid Comprehensive humanoid with 18 AX-12+ servos, which have been replaced in custom version -->
    <!--    When using this URDF, ensure a transform from odom to base_link is being published -->
<!--     <xacro:include filename="$(find bioloid_master)/urdf/bioloid_std_parts.xacro"/> -->
        <xacro:include filename="bioloid_torso_std_part.xacro"/>
<!-- 2) Custom parts of modified Bioloid Comprehensive humanoid, with CM-5 replaced by PC and custom sensors -->
    <!--    When using this URDF, ensure a transform from odom to imu_link is being published -->
<!--
     <xacro:include filename="$(find bioloid_master)/urdf/bioloid_custom_parts.xacro"/> 
-->

    <!-- Import gazebo_ros_control plugin -->
    <xacro:include filename="bioloid_without_legs_gazebo.xacro"/>

    <!-- <xacro:macro name="shoulder_swing_joint" params="prefix reflect x"> -->
    <xacro:macro name="shoulder_swing_joint" params="prefix reflect">
        <joint name="${prefix}_shoulder_swing_joint" type="revolute">
            <parent link="torso_link"/>
            <!-- <origin xyz="${reflect*(-0.032)} 0 0" rpy="0 ${reflect*(-pi/2)} ${x*pi}"/> -->
            <origin xyz="${reflect*(-0.032)} 0 0" rpy="${reflect*(-pi/2)} 0 ${reflect*(pi/2)}"/>
            <xacro:z_rot/>
            <child link="${prefix}_shoulder_swing_link"/>
            <limit lower="${-pi}" upper="${pi}" effort="10" velocity="1.0"/>
        </joint>
    </xacro:macro>
    <xacro:macro name="shoulder_lateral_joint" params="prefix reflect">
        <joint name="${prefix}_shoulder_lateral_joint" type="revolute">
            <parent link="${prefix}_shoulder_swing_link"/>
            <!-- <origin xyz="${reflect*(-0.014)} 0 ${reflect*0.044}" rpy="${-pi/2} 0 0"/> -->
            <origin xyz="${reflect*(-0.014)} 0 ${reflect*0.044}" rpy="${-pi/2} ${pi/2} 0"/>
            <xacro:z_rot/>
            <child link="${prefix}_upper_arm_link"/>
            <limit lower="${-pi}" upper="${pi}" effort="10" velocity="1.0"/>
        </joint>
    </xacro:macro>
    <xacro:macro name="elbow_joint" params="prefix reflect">
        <joint name="${prefix}_elbow_joint" type="revolute">
            <parent link="${prefix}_upper_arm_link"/>
            <origin xyz="${reflect*-0.0675} 0 0" rpy="0 0 0"/>
            <xacro:z_rot/>
            <child link="${prefix}_lower_arm_link"/>
            <limit lower="${-1.77}" upper="${1.77}" effort="10" velocity="1.0"/>
        </joint>
    </xacro:macro>
    <xacro:macro name="hand_joint" params="prefix">
        <joint name="${prefix}_hand_joint" type="fixed">
            <parent link="${prefix}_lower_arm_link"/>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <child link="${prefix}_hand_link"/>
        </joint>
    </xacro:macro>

    <!-- World -->
    <!-- Used for fixing robot to Gazebo 'base_link' -->

    <link name="world"/>

      <joint name="fixed" type="fixed">
        <parent link="world"/>
        <child link="base_link"/>
        <origin xyz="0 0 0.3" rpy="0 0 0"/>
      </joint>


    <!-- Base -->
    <joint name="base_to_torso_joint" type="fixed">
        <parent link="base_link"/>
        <origin xyz="0 0 0" rpy="0 0 ${pi/2}"/>
        <child link="torso_link"/>
    </joint>
    
    <link name="base_link">
    </link>
    
    <!-- Neck Joint-->
  <joint
    name="head_yaw_joint"
    type="revolute">
    <origin
      xyz="0 0 0.015"
      rpy="0 0 0" />
    <parent
      link="base_link" />
    <child
      link="neck_link" />
    <limit lower="${-pi}" upper="${pi}" effort="10" velocity="1.0"/>
    <axis
      xyz="0 0 1" />
    <limit
      effort="0"
      velocity="0" />
  </joint>

  <!-- Neck Link-->
    <link
    name="neck_link">
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://bioloid_example/mesh/neck_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.1 0.1 0.1 1" />
      </material>
    </visual>
  </link>
    
    <!-- Head Link-->
  <link
    name="head_link">
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="package://bioloid_example/mesh/head_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.5 0.5 0.5 1" />
      </material>
    </visual>
  </link>
  
  <!-- Head Joint-->
  <joint
    name="head_pitch_joint"
    type="revolute">
    <origin
      xyz="0.008 0 0.053"
      rpy="0 0 0" />
    <parent
      link="neck_link" />
    <child
      link="head_link" />
      <limit lower="-1.3" upper="1.4" effort="10" velocity="1.0"/>
    <axis
      xyz="0 1 0" />
    <limit
      effort="0"
      velocity="0" />
  </joint>
    
    
    <!-- Torso to right shoulder -->
    <xacro:shoulder_swing_joint prefix="right" reflect="1"/>
    <link name="right_shoulder_swing_link">
        <visual>
            <xacro:default_geometry cad_file="FP04-F1"/>
            <origin xyz="-0.014 0 0.044" rpy="0 ${pi} ${-pi/2}"/>
        </visual>
        <collision name="right_shoulder_collision">
            <origin xyz="-0.006 0 0.0375"/>
            <geometry>
                <box size="0.039 0.052 0.036"/>
            </geometry>
        </collision>
        <inertial>
            <origin xyz="0 0 0"/>
            <mass value="1.0"/>
            <inertia  ixx="0.001" ixy="0.0"  ixz="0.0"  iyy="0.001"  iyz="0.0"  izz="0.001"/>
        </inertial>
    </link>

    <!-- Torso to left shoulder -->
    <xacro:shoulder_swing_joint prefix="left" reflect="-1"/>
    <link name="left_shoulder_swing_link">
        <visual>
            <xacro:default_geometry cad_file="FP04-F1"/>
            <origin xyz="0.014 0 -0.044" rpy="0 0 ${pi/2}"/>
        </visual>
        <collision name="left_shoulder_collision">
            <origin xyz="0.006 0 -0.0375"/>
            <geometry>
                <box size="0.039 0.052 0.036"/>
            </geometry>
        </collision>
        <inertial>
            <origin xyz="0 0 0"/>
            <mass value="1.0"/>
            <inertia  ixx="0.001" ixy="0.0"  ixz="0.0"  iyy="0.001"  iyz="0.0"  izz="0.001"/>
        </inertial>
    </link>

    <!-- Right shoulder to upper arm -->
    <xacro:shoulder_lateral_joint prefix="right" reflect="1"/>
    <link name="right_upper_arm_link">
        <visual>
            <xacro:default_geometry cad_file="AX-12"/>
            <origin xyz="0 0 0" rpy="${pi} 0 ${pi/2}"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="BPF-BU"/>
            <origin xyz="0 0 0.0185" rpy="${pi/2} 0 0"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="BPF-WA"/>
            <origin xyz="0 0 0.019" rpy="${pi/2} 0 0"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="FP04-F3"/>
            <origin xyz="-0.0415 0 0" rpy="0 0 ${-pi/2}"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="FP04-F2"/>
            <origin xyz="-0.0675 0 0" rpy="0 ${pi/2} 0"/>
        </visual>
        <collision name="right_upper arm_collision_proximal">
            <origin xyz="-0.0148 0 0"/>
            <geometry>
                <box size="0.054 0.033 0.044"/>
            </geometry>
        </collision>
        <collision name="right_upper arm_collision_distal">
            <origin xyz="-0.0555 0 0"/>
            <geometry>
                <box size="0.047 0.033 0.052"/>
            </geometry>
        </collision>
        <inertial>
            <origin xyz="0 0 0"/>
            <mass value="1.0"/>
            <inertia  ixx="0.001" ixy="0.0"  ixz="0.0"  iyy="0.001"  iyz="0.0"  izz="0.001"/>
        </inertial>
    </link>

    <!-- Left shoulder to upper arm -->
    <xacro:shoulder_lateral_joint prefix="left" reflect="-1"/>
    <link name="left_upper_arm_link">
        <visual>
            <xacro:default_geometry cad_file="AX-12"/>
            <origin xyz="0 0 0" rpy="0 0 ${pi/2}"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="BPF-BU"/>
            <origin xyz="0 0 -0.0185" rpy="${-pi/2} 0 0"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="BPF-WA"/>
            <origin xyz="0 0 -0.019" rpy="${-pi/2} 0 0"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="FP04-F3"/>
            <origin xyz="0.0415 0 0" rpy="0 0 ${pi/2}"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="FP04-F2"/>
            <origin xyz="0.0675 0 0" rpy="0 ${-pi/2} 0"/>
        </visual>
        <collision name="left_upper arm_collision_proximal">
            <origin xyz="0.0148 0 0"/>
            <geometry>
                <box size="0.054 0.033 0.044"/>
            </geometry>
        </collision>
        <collision name="left_upper arm_collision_distal">
            <origin xyz="0.0555 0 0"/>
            <geometry>
                <box size="0.047 0.033 0.052"/>
            </geometry>
        </collision>
        <inertial>
            <origin xyz="0 0 0"/>
            <mass value="1.0"/>
            <inertia  ixx="0.001" ixy="0.0"  ixz="0.0"  iyy="0.001"  iyz="0.0"  izz="0.001"/>
        </inertial>
    </link>

    <!-- Right upper arm to lower arm -->
    <xacro:elbow_joint prefix="right" reflect="1"/>
    <link name="right_lower_arm_link">
        <!-- Motor 5 -->
        <visual>
            <xacro:default_geometry cad_file="AX-12"/>
            <origin xyz="0 0 0" rpy="${pi} 0 ${pi/2}"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="BPF-BU"/>
            <origin xyz="0 0 0.0185" rpy="${pi/2} 0 0"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="BPF-WA"/>
            <origin xyz="0 0 0.019" rpy="${pi/2} 0 0"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="FP04-F3"/>
            <origin xyz="-0.0415 0 0" rpy="0 0 ${-pi/2}"/>
        </visual>
        <collision name="right_lower_arm_collision">
            <origin xyz="-0.0135 0 0"/>
            <geometry>
                <box size="0.051 0.033 0.044"/>
            </geometry>
        </collision>
        <inertial>
            <origin xyz="0 0 0"/>
            <mass value="1.0"/>
            <inertia  ixx="0.001" ixy="0.0"  ixz="0.0"  iyy="0.001"  iyz="0.0"  izz="0.001"/>
        </inertial>
    </link>

    <!-- Left upper arm to lower arm -->
    <xacro:elbow_joint prefix="left" reflect="-1"/>
    <link name="left_lower_arm_link">
        <!-- Motor 6 -->
        <visual>
            <xacro:default_geometry cad_file="AX-12"/>
            <origin xyz="0 0 0" rpy="0 0 ${pi/2}"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="BPF-BU"/>
            <origin xyz="0 0 -0.0185" rpy="${-pi/2} 0 0"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="BPF-WA"/>
            <origin xyz="0 0 -0.019" rpy="${-pi/2} 0 0"/>
        </visual>
        <visual>
            <xacro:default_geometry cad_file="FP04-F3"/>
            <origin xyz="0.0415 0 0" rpy="0 0 ${pi/2}"/>
        </visual>
        <collision name="left_lower_arm_collision">
            <origin xyz="0.017 0 0"/>
            <geometry>
                <box size="0.058 0.033 0.044"/>
            </geometry>
        </collision>
        <inertial>
            <origin xyz="0 0 0"/>
            <mass value="1.0"/>
            <inertia  ixx="0.001" ixy="0.0"  ixz="0.0"  iyy="0.001"  iyz="0.0"  izz="0.001"/>
        </inertial>
    </link>

    <!-- Right lower arm to hand -->
    <xacro:hand_joint prefix="right"/>
    <link name="right_hand_link">
        <visual>
            <xacro:default_geometry cad_file="FP04-F11"/>
            <origin xyz="-0.0415 0 0" rpy="${-pi} ${pi/2} 0"/>
        </visual>
        <collision name="right_palm_collision">
            <origin xyz="-0.039 0 0"/>
            <geometry>
                <box size="0.0131 0.0265 0.044"/>
            </geometry>
        </collision>
        <collision name="right_fingers_collision">
            <origin xyz="-0.0701 -0.0077 0"/>
            <geometry>
                <box size="0.0575 0.0175 0.035"/>
            </geometry>
        </collision>
        <inertial>
            <origin xyz="0 0 0"/>
            <mass value="1.0"/>
            <inertia  ixx="0.001" ixy="0.0"  ixz="0.0"  iyy="0.001"  iyz="0.0"  izz="0.001"/>
        </inertial>
    </link>

    <!-- Left lower arm to hand -->
    <xacro:hand_joint prefix="left"/>
    <link name="left_hand_link">
        <visual>
            <xacro:default_geometry cad_file="FP04-F11"/>
            <origin xyz="0.0415 0 0" rpy="0 ${pi/2} 0"/>
        </visual>
        <collision name="left_palm_collision">
            <origin xyz="0.039 0 0"/>
            <geometry>
                <box size="0.0131 0.0265 0.044"/>
            </geometry>
        </collision>
        <collision name="left_fingers_collision">
            <origin xyz="0.0701 0.0077 0"/>
            <geometry>
                <box size="0.0575 0.0175 0.035"/>
            </geometry>
        </collision>
        <inertial>
            <origin xyz="0 0 0"/>
            <mass value="1.0"/>
            <inertia  ixx="0.001" ixy="0.0"  ixz="0.0"  iyy="0.001"  iyz="0.0"  izz="0.001"/>
        </inertial>
    </link>

</robot>
